### DISCUSSION TITLE
PorousFlow multyphase - not converge due to DIVERGED_PC_FAILED

### DISCUSSION THREAD

**Discussion by wangshengwa** [idaholab/moose#22251] [Q&A Modules: General] [answered]
Hi all,
I have a system of two layers; and aquitard that confines and aquifer. A well extract water from the aquitard. I would like, in a first step, to desaturate the aquitard and after that, in a second step,  inject a gas fluid. However, the input file with two phases does not converge, even though the model is only extracting water in the first step.

Looks like the problem comes from the `Preconditioning` system. I get 

`  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to SUBPC_ERROR 
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge
`

Looks like activating `hypre boomeramg` converges, but to a wrong result and diverges after a while.

A previous issue suggest that this happens because a variable is declare but not used [https://github.com/idaholab/moose/issues/18893](url) . In my case the gas pressure is not used in the first step.

Has anyone seen something similar before?

You can find my codes here: https://github.com/wangshengwa/PorousFlow-stuff/tree/main/2pf

Cheers and thanks in advance
Wang

**Comment by cpgr** [comment, marked as answer]
You might benefit from using the `s_scale` parameter in the capillary pressure function to avoid the derivative blowing up in the fully-saturated initial step - see https://mooseframework.inl.gov/modules/porous_flow/nonlinear_convergence_problems.html and https://mooseframework.inl.gov/modules/porous_flow/capillary_pressure.html

Also, you probably want to include some `PorousFlowMassTimeDerivative` kernels to make sure that the fluid masses balance.

**Reply by wangshengwa** [reply]
Hi @cpgr . Thanks for your help is really nice. I forgot to add `PorousFlowMassTimeDerivative`, I was doing steady state tests. Now I fix that. Thanks.

I also studied the `s_scale` factor. However, for me is unclear what value should it take. Looks like different values of `s_scale` take the model to converge to different results. However, non of them seems correct.

Here, you will find a comparison of the "first stage" of my 2 phase model with a model that uses one phase and actions If you run them, you will see that the extraction rate in both models is different.

codes: https://github.com/wangshengwa/PorousFlow-stuff/tree/main/2pf

Can you advise how to proceed?
Cheers and thanks!
Wang

**Reply by cpgr** [reply]
Oh, I think I see the difference. Your single-phase unsaturated case (`action.i`) withdraws mass from the aquifer, but doesn't replace it with anything - so water saturation drops below 1, but the remaining pore space is empty (there is no gas phase). 

Your two-phase case (`kernels2p.i`) withdraws mass from phase 0, so that the saturation decreases below 1. In this case, however, there is another phase (phase 1) which fills the empty pore space  automatically (phase 1). This is different from the unsaturated case above. This case isn't unsaturated!

So I guess it depends on what you want to achieve here.

You could make your two-phase case be identical to the unsaturated case by setting the density of the phase 1 fluid equal to zero (this is a bit of a hack)
```
    [the_simple_fluid2]
      type = SimpleFluidProperties
      density0 = 0
    []
```
and you should now get the same results as `action.i` (note: you should make the nl tolerances identical to those in `action.i`).

**Reply by cpgr** [reply]
And the following is what `action.i` really is doing (with the kernels explicitly included)
```
[Mesh]
  [the_mesh]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 5
    ny = 100
    xmin = 0
    xmax = 5
    ymin = 0
    ymax = 100
  []
  [aquitard]
    type = SubdomainBoundingBoxGenerator
    block_id = 1
    bottom_left = '0 50 0'
    top_right = '5 100 0'
    input = the_mesh
  []
  [aquifer]
    type = SubdomainBoundingBoxGenerator
    block_id = 2
    bottom_left = '0 0 0'
    top_right = '5 50 0'
    input = aquitard
  []
[]

[GlobalParams]
  PorousFlowDictator = dictator
  biot_coefficient = 1.0
[]

[Variables]
  [ppwater]
  []
[]

[ICs]
  [ppwater]
    type = FunctionIC
    variable = ppwater
    function = '10000*(100-y)'
  []
[]


[BCs]
  [ppwater_top]
    type = FunctionDirichletBC
    variable = ppwater
    function = 0
    boundary = top
  []
[]

[Modules]
  [FluidProperties]
    [the_simple_fluid]
      type = SimpleFluidProperties
    []
  []
[]

[AuxVariables]
  [swater]
    family = MONOMIAL
    order = FIRST
  []
[]

[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'ppwater'
    number_fluid_phases = 1
    number_fluid_components = 1
  []
  [pc]
    type = PorousFlowCapillaryPressureVG
    m = 0.6
    alpha = 1e-5
    sat_lr = 0.0
  []
  [produced_mass]
    type = PorousFlowSumQuantity
  []
[]

[Kernels]
  [mass0]
    type = PorousFlowMassTimeDerivative
    fluid_component = 0
    variable = ppwater
  []
  [flux0]
    type = PorousFlowAdvectiveFlux
    fluid_component = 0
    variable = ppwater
    gravity = '0 -10 0'
  []
[]

[AuxKernels]
  [swater]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 0
    variable = swater
  []
[]

[Materials]
  [temperature]
    type = PorousFlowTemperature
  []
  [simple_fluid0]
    type = PorousFlowSingleComponentFluid
    fp = the_simple_fluid
    phase = 0
  []
  [ppss]
    type = PorousFlow1PhaseP
    porepressure = ppwater
    capillary_pressure = pc
  []
  [massfrac]
    type = PorousFlowMassFraction
  []

  [porosity_clay]
    type = PorousFlowPorosity
    porosity_zero = 0.3
    block = 1
  []
  [permeability_clay]
    type = PorousFlowPermeabilityConst
    permeability = '1E-20 0 0   0 1E-20 0   0 0 1E-20'
    block = 1
  []
  [porosity_rock]
    type = PorousFlowPorosity
    porosity_zero = 0.05
    block = 2
  []
  [permeability_rock]
    type = PorousFlowPermeabilityConst
    permeability = '1E-13 0 0   0 1E-13 0   0 0 1E-14'
    block = 2
  []
  [relperm_water]
    type = PorousFlowRelativePermeabilityCorey
    n = 3
    phase = 0
  []
[]

[DiracKernels]
   [produce]
     type = PorousFlowPeacemanBorehole
     variable = ppwater
     SumQuantityUO = produced_mass
     mass_fraction_component = 0
     point_file = action.bh
     bottom_p_or_t = 1E-8
     unit_weight = '0 0 0'
     use_mobility = true
     character = 1
   []
[]

[Postprocessors]
  [bh_report]
    type = PorousFlowPlotQuantity
    uo = produced_mass
    execute_on = 'initial timestep_end'
  []
  [fluid_mass]
    type = PorousFlowFluidMass
    execute_on = 'initial timestep_end'
  []
[]


[Preconditioning]
  active = basic
  [basic]
    type = SMP
    full = true
    petsc_options = '-ksp_diagonal_scale -ksp_diagonal_scale_fix'
    petsc_options_iname = '-pc_type -sub_pc_type -sub_pc_factor_shift_type -pc_asm_overlap'
    petsc_options_value = ' asm      lu           NONZERO                   2'
  []
  [lu]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = ' lu       mumps'
  []
  [moose]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
    petsc_options_value = 'hypre boomeramg 500'
  []
[]

[Executioner]
  type = Transient
  solve_type = Newton
  start_time = -3600
  dt = 3600
  end_time = 172800
  nl_abs_tol = 1E-12
  nl_rel_tol = 1e-08
[]

[Outputs]
  print_linear_residuals = false
  exodus = true
[]
```

**Reply by wangshengwa** [reply]
Thanks! Setting density of the phase 1 to zero works! However, I would like to in a first stage desaturate the medium of the phase 0 (leave it empty) and in a second stage inject the phase 1.

If I understood correctly, by setting `density0 = 0` we are still replacing the empty space with the phase 1 is just that this phase has no mass. But in the second stage I would like to inject fluid with mass, how can I solve this?

Cheers!
Wang

**Reply by WilkAndy** [reply]
This is just an opinion, not necessarily fact.... i think that your concept of withdrawing water and then replacing the vacuum with air is flawed.  It will be difficult to run PorousFlow like this because it is unphysical (an aquifer never contains a vacuum - always air or some other gas).  If i were you, i'd attempt to reproduce the physical situation with your model.  Having said this, PorousFlow was used to simulate water+vacuum in a carbon-fibre composite, where there truly was a vacuum, so it's not impossible, but care had to be taken with the capillary curve.  Almost definitely you are getting problems from the end-points of the capillary pressure curve, and probably the boundary conditions in your present setup (eg, you are extracting a fluid component from a region where it doesn't exist).

**Reply by wangshengwa** [reply]
Hi and thanks for your reply. Indeed my approach seems wrong. Thanks for your comment. 
In other words, only makes physical sense to keep the aquifer saturated at all times. 

After some testing, I noticed that as soon as the aquifer desaturates MOOSE has problems converging.
Thanks!

**Comment by friedmud** [comment]
My suggestion would be to either use the `MultiApp` system plus the `FullSolveMultiApp` to do the initial phase... or to do this as a restart problem.

https://mooseframework.inl.gov/source/multiapps/FullSolveMultiApp.html
https://mooseframework.inl.gov/application_usage/restart_recover.html

**Reply by wangshengwa** [reply]
Hi @friedmud I will try this tomorrow and come back to you. Thanks!
Wang
